p2p/discovery.go

```diff
@@ -0,0 +1,204 @@
+package p2p
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/multiformats/go-multiaddr"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+const DiscoveryProtocol = "/tss/discovery/1.0.0"
+
+const MaxGossipConcurrency = 50
+
+var GossipInterval = 30 * time.Second
+
+type PeerDiscovery struct {
+	host           host.Host
+	knownPeers     map[peer.ID]peer.AddrInfo
+	bootstrapPeers []peer.AddrInfo
+	mu             sync.RWMutex
+	logger         zerolog.Logger
+	closeChan      chan struct{}
+}
+
+func NewPeerDiscovery(h host.Host, bootstrapPeers []peer.AddrInfo) *PeerDiscovery {
+	pd := &PeerDiscovery{
+		host:           h,
+		knownPeers:     make(map[peer.ID]peer.AddrInfo),
+		bootstrapPeers: bootstrapPeers,
+		logger:         log.With().Str("module", "peer-discovery").Logger(),
+		closeChan:      make(chan struct{}),
+	}
+
+	// Set up discovery protocol handler
+	h.SetStreamHandler(DiscoveryProtocol, pd.handleDiscovery)
+
+	return pd
+}
+
+// Start begins the discovery process
+func (pd *PeerDiscovery) Start(ctx context.Context) {
+	pd.logger.Info().Msgf("Starting peer discovery with bootstrap peers: %v", pd.bootstrapPeers)
+	// Connect to bootstrap peers first
+	for _, pinfo := range pd.bootstrapPeers {
+		if err := pd.host.Connect(ctx, pinfo); err != nil {
+			pd.logger.Error().Err(err).Msgf("Failed to connect to bootstrap peer %s", pinfo.ID)
+			continue
+		}
+		pd.addPeer(pinfo)
+	}
+
+	//before periodic gossip, start two rounds of warmup; this is to ensure keygen/keysign unit test
+	// success where there might not be enough time for gossip to propagate before the keygen starts.
+	pd.gossipPeers(ctx)
+	time.Sleep(1 * time.Second)
+	pd.gossipPeers(ctx)
+	// Start periodic gossip
+	go pd.startGossip(ctx)
+}
+
+func (pd *PeerDiscovery) Stop() {
+	close(pd.closeChan)
+}
+
+// addPeer adds a peer to known peers
+func (pd *PeerDiscovery) addPeer(pinfo peer.AddrInfo) {
+	pd.mu.Lock()
+	defer pd.mu.Unlock()
+
+	if pinfo.ID == pd.host.ID() {
+		return // Don't add ourselves
+	}
+	pd.knownPeers[pinfo.ID] = pinfo
```

Comment:
If we aren't going to filter private addresses, we'd need to allow storing multiple addresses per peer right? Because one peer could gossip a private address that's not connectable?

Reply:
dae1f21c4c3b71ab14bdcc833c815fa41c9a8433